# **Project:** Cart Pole

## **Problem Description**

The inverted pendulum problem (cart-pole problem) is often used as one of the fundamental examples in reinforcement learning. It consists of a rod attached by an unmovable joint to a cart that moves linearly along a track, without friction. The center of mass of the rod is located above its pivot point. The goal is to maintain the system balanced by applying appropriate forces to the cart itself. This physical system, due to its nonlinear dynamics and unstable operating point, presents a challenge in the control domain. In this regard, the idea of this final project is to apply one of the reinforcement learning methods to control such a physical system. The schematic of this problem setup is shown in Figure 1.

<div align="center">

![alt text](cart_pole-1.svg)

</div>

The inverted pendulum can be modeled by a system of nonlinear equations:

$$\ddot{\theta} = \frac{Mg \sin \theta - \cos \theta(F + ml \dot{\theta}^2 \sin \theta)}{(1 + k)Ml - ml \cos^2 \theta}$$

$$\ddot{x} = \frac{mg \sin \theta \cos \theta - (1 + k)(F + ml \dot{\theta}^2 \sin \theta)}{m \cos^2 \theta - (1 + k)M}$$

*Equations are taken from https://sharpneat.sourceforge.io/research/cart-pole/cart-pole-equations.html*

**Parameters:**
- **m** - mass of the rod (100g)
- **M** - total mass of rod and cart (1.1kg)  
- **l** - length of the rod (0.5m)
- **g** - gravitational constant (no suggested value)
- **θ** - angle between rod and vertical axis
- **x** - position
- **F** - force (F ∈ [-10N, 10N])
- **k** - constant related to rod's moment of inertia (k = 0; k = 1/3; k = 1)

*Values in parentheses are suggested values you can use in implementation.*

In compact form, the model can also be written as:

$$\ddot{\theta} = f_\theta(x, \dot{x}, \theta, \dot{\theta}, F)$$
$$\ddot{x} = f_x(x, \dot{x}, \theta, \dot{\theta}, F)$$

where $f_\theta$ and $f_x$ are nonlinear functions.

--- 

**Converting Mathematical Model to MDP**

To control such a model using methods we've learned in this course, we need to convert it to a suitable form. Although Markov Decision Processes can be generalized to cases where time flows continuously, we won't deal with that here. Instead, we'll convert the given model to a form equivalent to a Markov Decision Process: the next state and reward are functions of the current state and applied action, i.e.:

$$s^+ = f(s, a)$$
$$r = h(s, a)$$

Although it may not be obvious, the first step is to reduce the order of the system given by equations (2). By introducing substitutions:

$$z_1 = x$$

$$z_2 = \dot{x}$$

$$z_3 = \theta$$

$$z_4 = \dot{\theta}$$

and differentiating the left side of each equation with respect to time, we get a mathematical model in state space form:

$$\dot{z_1} = z_2$$

$$\dot{z_2} = f_x(z, F)$$

$$\dot{z_3} = z_4$$

$$\dot{z_4} = f_\theta(z, F)$$

where $z = (z_1, z_2, z_3, z_4)$. In other words, all state variables are enumerated in z.

The second step is discretization. For this task, we'll choose Euler discretization, i.e., numerical integration with left rectangles, because for nonlinear models it's sometimes simply not possible to apply anything else. By introducing the substitution $\dot{z_i} = \frac{1}{T}(z_i^+ - z_i)$, i = 1,2,3,4, which corresponds to the first derivative approximation in forward differencing, we get the equations:

$$z_1^+ = z_1 + T z_2$$

$$z_2^+ = z_2 + T f_x(z, F)$$

$$z_3^+ = z_3 + T z_4$$

$$z_4^+ = z_4 + T f_\theta(z, F)$$

The physical system is now converted to MDP form, with the detail that we now have a state vector $z = (z_1, z_2, z_3, z_4)$. On the other hand, action $a$ is nothing other than the force $F$ applied to the cart at a given moment in time. The action is also discrete and bounded.

---

**Reward Formation**

The question of reward formation is essential, i.e., the function $h$ in equation $r = h(s, a)$ of the deterministic MDP. In this section, we offer several suggestions for reward function formation:

**Option 1:**
$$r = \begin{cases} 
|\theta|^\alpha & \text{if } \theta < 60° \\
-1000 & \text{if } \theta \geq 60°
\end{cases}$$

**Option 2:**
$$r = \begin{cases}
0 & \text{if } \theta = ±90° \\
-100 & \text{if } \theta \neq ±90°
\end{cases}$$

**Option 3:**
$$r = \begin{cases}
1 & \text{if } \theta = ±90° \\
0 & \text{if } \theta \neq ±90°
\end{cases}$$

The last suggestion is given following the predefined Gymnasium environment for the inverted pendulum. At the following hyperlinks, you can find general instructions for using Gymnasium environments, training agents, and using the cart-pole environment itself.

**Useful Links:**
- [Using environments](https://gymnasium.farama.org/)
- [Agent training example](https://stable-baselines3.readthedocs.io/)
- [Cart-pole environment](https://gymnasium.farama.org/environments/classic_control/cart_pole/)

We note that using the predefined environment is not mandatory; we only suggest it as an option. Although we would ideally greatly appreciate you implementing the environment and solution yourself, we offer Gymnasium as a (potentially) facilitating circumstance. In this regard, the decision whether to use it will not affect the number of points earned on this final task.

##### **Note:** Version on Serbian

**Opis problema**

Problem inverznog klatna (eng. **cart-pole problem**) često se koristi kao jedan od osnovih primera u učenju potkrepljenjem. To je šipka je pričvršćena nepokretnim zglobom za kolica koja se kreću pravolinijski duž pruge, bez trenja. Centar mase šipke nalazi se iznad njene pivot tačke. Cilj je održavati sistem balansiranim, i to primenom adekvatnih sila na sama kolica. Ovakav fizički sistem, zbog svoje nelinearne dinamike i nestabilne radne tačke, predstavlja izazov u domenu upravljanja. S tim u vezi, ideja ovog završnog domaćeg zadatka je da se primeni neka od metoda učenja potkrepljenjem za upravljanje ovakvim fizičkim sistemom.

<div align="center">

![alt text](cart_pole-1.svg)

</div>

**Slika 1:** Prikaz modela inverznog klatna.

---

**Matematički model**

Inverzno klatno može da se modeluje sistemom nelinearnih jednačina:

$$\ddot{\theta} = \frac{Mg \sin \theta - \cos \theta(F + ml \dot{\theta}^2 \sin \theta)}{(1 + k)Ml - ml \cos^2 \theta}$$

$$\ddot{x} = \frac{mg \sin \theta \cos \theta - (1 + k)(F + ml \dot{\theta}^2 \sin \theta)}{m \cos^2 \theta - (1 + k)M}$$

*Jednačine su preuzete sa https://sharpneat.sourceforge.io/research/cart-pole/cart-pole-equations.html*

**Parametri modela:**
- **m** - masa štapa (100g)
- **M** - ukupna masa štapa i kolica (1.1kg)  
- **l** - dužina šipke (0.5m)
- **g** - gravitaciona konstanta (nema predložene vrednosti)
- **θ** - ugao između šipke i vertikalne ose
- **x** - pozicija
- **F** - sila (F ∈ [-10N, 10N])
- **k** - konstanta vezana za moment inercije štapa (k = 0; k = 1/3; k = 1)

U sažetoj formi, model možemo zapisati i u obliku:

$$\ddot{\theta} = f_\theta (x, \dot{x}, \theta, \dot{\theta}, F)$$
$$\ddot{x} = f_x(x, \dot{x}, \theta, \dot{\theta}, F)$$

pri čemu su $f_\theta$ i $f_x$ nelinearne funkcije.

---

**Prevođenje matematičkog modela u MPO**

Da bismo ovakvim modelom mogli da upravljamo nekim od metoda sa kojima smo se upoznali na ovom predmetu, potrebno je da ga prevedemo u pogodnu formu. Premda se Markovljevi procesi odlučivanja mogu uopštiti i na slučajeve kada vreme teče kontinualno, mi se time ovde nećemo baviti. Umesto toga, dati model ćemo prevesti u formu koja je ekvivalentna Markovljevom procesu odlučivanja: naredno stanje i nagrada su funkcije tekućeg stanja i primenjene akcije, odnosno:

$$s^+ = f(s, a)$$
$$r = h(s, a)$$

**Korak 1: Snižavanje reda sistema**

Uvođenjem smena:
$$z_1 = x$$
$$z_2 = \dot{x}$$
$$z_3 = \theta$$
$$z_4 = \dot{\theta}$$

i diferenciranjem po vremenu leve strane svake jednačine, dobijamo matematički model u prostoru stanja u obliku:

$$\dot{z_1} = z_2$$
$$\dot{z_2} = f_x(z, F)$$
$$\dot{z_3} = z_4$$
$$\dot{z_4} = f_\theta(z, F)$$

pri čemu je $z = (z_1, z_2, z_3, z_4)$.

**Korak 2: Diskretizacija (Euler metoda)**

Za potrebe ovog zadatka, izabraćemo Euler diskretizaciju, tj numeričku integraciju levim pravougaonicima. Uvođenjem smene $\dot{z_i} = \frac{1}{T}(z_i^+ - z_i)$, dobijamo jednačine:

$$z_1^+ = z_1 + T z_2$$
$$z_2^+ = z_2 + T f_x(z, F)$$
$$z_3^+ = z_3 + T z_4$$
$$z_4^+ = z_4 + T f_\theta(z, F)$$

Fizički sistem je sada preveden u formu MPO, sa detaljem da sada imamo **vektor stanja** $z = (z_1, z_2, z_3, z_4)$. S druge strane **akcija** $a$ je ništa drugo nego sila $F$ koja se u datom trenutku vremena primenjuje na kolica. Akcija je takođe diskretna i ograničena veličina.

---

**Formiranje nagrade**

Neizostavno je i pitanje načina formiranja nagrade, odnosno funkcije $h$ u jednačini $r = h(s, a)$ determinističkog MPO. Evo nekoliko predloga za način formiranja funkcije nagrade:

**1. Nagrada bazirana na uglu (sa eksponentom):**
$$r = \begin{cases}
|\theta|^\alpha & \text{ako } \theta < 60° \\
-1000 & \text{ako } \theta \geq 60°
\end{cases}$$

**2. Binarna nagrada (klasična):**
$$r = \begin{cases}
0 & \text{ako } \theta = ±90° \\
-100 & \text{ako } \theta \neq ±90°
\end{cases}$$

**3. Pozitivna nagrada (Gymnasium stil):**
$$r = \begin{cases}
1 & \text{ako } \theta = ±90° \\
0 & \text{ako } \theta \neq ±90°
\end{cases}$$

Poslednji predlog je dat po uzoru na predefinisano **Gymnasium** okruženje za inverzno klatno.

**Korisni linkovi za Gymnasium:**
- [Korišćenje okruženja](https://gymnasium.farama.org/)
- [Primer treniranja agenta](https://gymnasium.farama.org/tutorials/)
- [Cart-pole okruženje](https://gymnasium.farama.org/environments/classic_control/cart_pole/)

**Napomena:** Korišćenje predefinisanog okruženja nije obavezno, ovde ga samo predlažemo kao opciju. Iako bismo u idealnom slučaju izuzetno cenili da sami implementirate okruženje i rešenje, Gymnasium nudimo kao (potencijalno) olakšavajuću okolnost. S tim u vezi, odluka o tome da li ga koristite neće uticati na broj ostvarenih bodova na ovom završnom zadatku.

---

**Zadaci za samostalan rad**

1. **Diskretizacija matematičkog modela u prostoru stanja**
   - U slučaju da koristite Gymnasium, ovaj korak nije potreban.

2. **Definisanje prostora akcija**
   - Napominjemo da prostor akcija treba da bude diskretan i ograničen.
   - U slučaju da koristite Gymnasium, prostor akcija je već predefinisan.

3. **Izbor načina formiranja nagrade**
   - Nije obavezno da izaberete neki od datih predloga
   - Slobodno osmislite originalan način formiranja nagrade.

4. **Treniranje agenta pomoću Q-Learning ili SARSA**

5. **Simulacija okruženja i prikaz dobijenih rezultata**
   - Ukoliko se ne služite Gymnasium okruženjem, nije obavezno da pravite nikakvu vizualizaciju.

## **Assignment**

Tasks for independent work:

1. **Perform discretization of the mathematical model in state space.** *If using Gymnasium, this step is not necessary.*

2. **Define the action space.** Note that the action space should be discrete and bounded. *If using Gymnasium, the action space is already predefined.*

3. **Choose a method for reward formation.** It's not mandatory to choose one of the given suggestions; feel free to devise an original way of forming rewards.

4. **Train an agent using Q-Learning or SARSA.**

5. **Simulate the environment and display the obtained results.** If you're not using the Gymnasium environment, visualization is not mandatory.

## **Solution:** Implemented Environment

This solution is implemented without using external libraries and it contains my implementation of Cart Pole problem environment.

#### Basic imports

In [181]:
import math
import random
import numpy as np
import matplotlib.pyplot as plt
from collections import defaultdict

### **Implement:** `Cart-Pole Model`

**Task 1:** **Perform discretization of the mathematical model in state space.** *If using Gymnasium, this step is not necessary.*

In [182]:
class CartPoleModel:
    """
    Physics model of Cart-Pole system.
    Simulates how the cart and pole move based on forces applied.
    """

    def __init__(self, T = 0.02, m = 0.1, M = 1.1, l = 0.5, g = 9.81, k = 1/3):
        """
        Initialize physics model of Cart-Pole system.
        
        Parameters:
            T: time step (seconds)
            m: mass of the pole (kg)
            M: mass of the cart (kg)
            l: length of the pole (meters)
            g: gravitational acceleration (m/s²)
            k: moment of inertia constant
        """
        self.T = T
        self.m = m
        self.M = M
        self.l = l
        self.g = g
        self.k = k

    def f_x(self, z, F):
        """
        Calculates horizontal acceleration of the cart by using one of given nonlinear equations.
        """
        _, _, theta, theta_dot = z
        sin_theta = math.sin(theta)
        cos_theta = math.cos(theta)
        numerator = (self.m * self.g * sin_theta * cos_theta - (1 + self.k) * (F + self.m * self.l * theta_dot**2 * sin_theta))
        denominator = (self.m * cos_theta**2 - (1 + self.k) * self.M)

        return numerator / denominator if abs(denominator) > 1e-10 else 0
    
    def f_theta(self, z, F):
        """
        Calculate angular acceleration of the pole by by using one of given nonlinear equations.
        """
        _, _, theta, theta_dot = z
        sin_theta = math.sin(theta)
        cos_theta = math.cos(theta)
        numerator = (self.M * self.g * sin_theta - cos_theta * (F + self.m * self.l * theta_dot**2 * sin_theta))
        denominator = ((1 + self.k) * self.M * self.l - self.m * self.l * cos_theta**2)

        return numerator / denominator if abs(denominator) > 1e-10 else 0

    def next_state(self, z, force):
        """
        Calculate next state using Euler integration.
        """
        z_next = np.zeros(4)

        z_next[0] = z[0] + self.T * z[1]                        # Update position
        z_next[1] = z[1] + self.T * self.f_x(z, force)          # Update velocity
        z_next[2] = z[2] + self.T * z[3]                        # Update angle
        z_next[3] = z[3] + self.T * self.f_theta(z, force)      # Update angle velocity

        return z_next

### **Implement:** `Cart-Pole State Discretizer`

In [183]:
class CartPoleStateDiscretizer:
    """
    Converts Cart-Pole continuous state space to discrete states (bins) for Q-Learning.
    So we take each state variable and split them in provided number of bins.
    """
    
    def __init__(self, bins=(10, 10, 10, 10), limits=None):
        """
        Initialize Cart-Pole state discretizer.

        Parameters:
            bins: number of bins for each state variable (x, x_dot, theta, theta_dot)
            limits: tuple of limits for each state variable
        """
        self.bins = bins

        if limits is not None:
            # Use provided limits for each state variable
            self.x_limit = limits[0]
            self.x_dot_limit = limits[1]
            self.theta_limit = limits[2]
            self.theta_dot_limit = limits[3]
        else:
            # Default limits for Cart-Pole problem
            self.x_limit = (-2.4, 2.4)
            self.x_dot_limit = (-3.0, 3.0)
            self.theta_limit = (-0.20944, 0.20944)
            self.theta_dot_limit = (-2.0, 2.0)

    def discretize_state(self, state):
        """
        Convert continuous state into discrete bins.
            np.clip() - If values is outside of range, set it on the boundary (limit value)
            np.digitize - Finds which bin each value belongs to, returns 1-indexed, so we 
                          subtract 1, to get 0-indexed)
        """
        state = tuple(state)                    
        x, x_dot, theta, theta_dot = state

        # Clip state variables to their limits
        x = np.clip(x, self.x_limit[0], self.x_limit[1])
        x_dot = np.clip(x_dot, self.x_dot_limit[0], self.x_dot_limit[1])
        theta = np.clip(theta, self.theta_limit[0], self.theta_limit[1])
        theta_dot = np.clip(theta_dot, self.theta_dot_limit[0], self.theta_dot_limit[1])

        # Create bin edges for each state variable 
        x_bins = np.linspace(self.x_limit[0], self.x_limit[1], self.bins[0] + 1)
        x_dot_bins = np.linspace(self.x_dot_limit[0], self.x_dot_limit[1], self.bins[1] + 1)
        theta_bins = np.linspace(self.theta_limit[0], self.theta_limit[1], self.bins[2] + 1)
        theta_dot_bins = np.linspace(self.theta_dot_limit[0], self.theta_dot_limit[1], self.bins[3] + 1)
        
        # Find which bin each state variable falls into (discretize each dimension)
        x_discrete = np.digitize(x, x_bins) - 1
        x_dot_discrete = np.digitize(x_dot, x_dot_bins) - 1
        theta_discrete = np.digitize(theta, theta_bins) - 1
        theta_dot_discrete = np.digitize(theta_dot, theta_dot_bins) - 1

        # Ensure indices (indexes) are within valid range (handle edge cases)
        x_discrete = max(0, min(x_discrete, self.bins[0] - 1))
        x_dot_discrete = max(0, min(x_dot_discrete, self.bins[1] - 1))
        theta_discrete = max(0, min(theta_discrete, self.bins[2] - 1))
        theta_dot_discrete = max(0, min(theta_dot_discrete, self.bins[3] - 1))

        return (x_discrete, x_dot_discrete, theta_discrete, theta_dot_discrete)

**Note:**
- `state = tuple(state)`: state needs to be hashable

### **Implement:** `Cart-Pole Action Space`

**Task 2:** **Define the action space.** Note that the action space should be discrete and bounded. *If using Gymnasium, the action space is already predefined.*

#### **Version 1.0**

In [184]:
class ActionSpace:
    """
    Defines action space with actions that agent can take.
    Agent take some action ---> Agent apply some force
    """

    def __init__(self, actions_type = 'binary', max_force: float = 10.0):
        """
        Initialize action space for Cart-Pole system.

        Parameters:
            actions_type: 'binary' | 'ternary' | 'multi'
            max_force: maximum force that can be applied
        """
        self.actions_type = actions_type
        self.max_force = max_force

        # Defined action space, discrete and bounded (with limits), based on chosen actions_type
        if actions_type == 'binary':
            self.action_num = 2
            self.actions = [-max_force, +max_force]
            self.action_names = ['Left (-10)', 'Right (+10)']
        
        elif actions_type == 'ternary':
            self.action_num = 3
            self.actions = [-max_force, 0.0, +max_force]
            self.action_names = ['Left (-10)', 'No Force (0)', 'Right (+10)']

        elif actions_type == 'multi':
            self.action_num = 5
            self.actions = [-max_force, -max_force/2, 0.0, +max_force/2, +max_force]
            self.action_names = ['Strong Left (-10)',
                                 'Weak Left (-10)',
                                 'No Force (0)',
                                 'Weak Right (+10)',
                                 'Strong Right (+10)']

    def get_random_action(self):
        """
        Get a random action index for exploration of environment.
        """
        return np.random.randint(0, self.action_num)
    
    def get_force(self, action_idx: int):
        """
        Convert a action index in force value and description.
        """
        if 0 <= action_idx < self.action_num:
            force = self.actions[action_idx]
            force_description = self.action_names[action_idx]
            return force, force_description
        else:
            raise ValueError(f"Invalid action index: {action_idx}")

### **Implement:** `Cart-Pole Reward Formation Functions`

**Task 3:** **Choose a method for reward formation.** It's not mandatory to choose one of the given suggestions; feel free to devise an original way of forming rewards.

#### **Version 1.0**

In [197]:
class RewardFunction:
    """
    Defines reward formation function that return reward agent received after taken action.
    """

    def __init__(self, reward_type = 'angle_based', alpha = 2.0, terminal_theta = 0.20944):
        """
        Initialize reward formation function for Cart-Pole system.

        Parameters:
            reward_type: 'angle_based' | 'binary' | 'gym_based'
            alpha: exponent parameter for first reward function
            terminal_theta: terminal angle that will flag that state is terminal (task failed)
        """
        self.reward_type = reward_type
        self.alpha = alpha
        self.terminal_theta = terminal_theta

    def angle_based_reward(self, theta: float):
        """
        Angle-based reward function. Rewards smaller deviations from vertical.
        """
        theta_degree = math.degrees(theta)
        if abs(theta_degree) < 60:
            return -abs(theta_degree)**self.alpha  # Based on angle between pole and vertical axis
        else:
            return -1000.0  # Penalty is pole cross angle limit

    def binary_reward(self, theta: float):
        """
        Binary reward function. Anti-reward (penalty) strategy.
        """
        if abs(theta) >= self.terminal_theta:
            return -100.0  # Penalty for terminal state
        else:
            return 0.0  # No reward for non-terminal states

    def gym_based_reward(self):
        """
        Gymnasium type binary reward function. Agent is rewarded if pole is not down.
        """
        return 1.0  # Reward for surviving


    def get_reward(self, theta: float = None):
        """
        Get reward by reward function of chosen reward type.
        Note: For gym_based, theta is not used.
        """
        if self.reward_type == 'angle_based':
            if theta is None:
                raise ValueError("Theta required for angle_based reward.")
            reward = self.angle_based_reward(theta)
        elif self.reward_type == 'binary':
            if theta is None:
                raise ValueError("Theta required for binary reward.")
            reward = self.binary_reward(theta)
        elif self.reward_type == 'gym_based':
            reward = self.gym_based_reward()  # No dependency on theta
        else:
            raise ValueError("Unknown reward function type.")

        return reward

### **Implement:** `Cart-Pole Q-Learning Agent`

**Task 4:** **Train an agent using Q-Learning or SARSA.**

In [198]:
class QLearningAgent:
    """
    Q-Learning agent for solving Cart-Pole problem.
    """

    def __init__(self, action_size, gamma = 0.95, epsilon = 1.0, epsilon_min = 0.01,
                 epsilon_decay = 0.995, alpha = 0.1):
        """
        Initialize Q-Learning agent.

        Parameters:
            action_size: ActionSpace size (number of actions)
            gamma: discount factor
            epsilon: initial exploration rate
            epsilon_min: minimum epsilon value
            epsilon_decay: decay rate for epsilon
            alpha: learning rate
        """
        self.action_size = action_size
        self.gamma = gamma
        self.epsilon = epsilon
        self.epsilon_min = epsilon_min
        self.epsilon_decay = epsilon_decay
        self.alpha = alpha

        self.q_table = defaultdict(lambda: np.zeros(action_size))

        self.total_rewards = []
        self.total_steps = []

    def get_action(self, state, training_active = True):
        """
        Choose action using epsilon-greedy policy.
        """
        if training_active and np.random.random() < self.epsilon:
            # Exploration: choose random action
            return np.random.randint(0, self.action_size)
        else:
            # Exploitation: choose best action based on Q-values
            return np.argmax(self.q_table[state])

    def decay_epsilon(self):
        """
        Decay (lovers) exploration rate after each episode.
        """
        self.epsilon = max(self.epsilon_min, self.epsilon * self.epsilon_decay)

    def update_q_table(self, state, action, reward, next_state, terminal):
        """
        Update Q-table using Q-Learning update formula.
        Q(s,a) <- Q(s,a) + alpha * (r + gamma*max_a' Q(s',a')*(1 - terminal) - Q(s,a))

        If terminal: td_target = reward (no bootstrap).
        """
        if terminal:
            td_target = reward
        else:
            best_next = np.max(self.q_table[next_state])
            td_target = reward + self.gamma * best_next

        td_error = td_target - self.q_table[state][action]

        self.q_table[state][action] += self.alpha * td_error

### **Implement:** `Train Q-Learning Agent Method`

In [199]:
def train_agent(environment, state_discretizer, agent, episodes_number, max_steps=500):
    """
    The main function for training the Q-learning agent.

    Parameters:
        environment: The environment of the cart and stick
        state_discretizer: State discretizer
        agent: Q-learning agent
        episodes_number: Number of training episodes
        max_steps: Maximum steps per episode
    """
    training_rewards = []
    training_steps = []

    for episode in range(episodes_number):
        # Reset environment to initial state
        state = environment.reset()
        state_discrete = state_discretizer.discretize_state(state)

        episode_reward = 0
        episode_steps = 0
        terminal = False

        # Run episode until terminal state or max steps reached
        while not terminal and episode_steps < max_steps:
            # Agent chooses action
            action = agent.get_action(state_discrete)

            # Execute action in environment
            next_state, reward, terminal = environment.step(action)

            # Discretize next state
            next_state_discrete = state_discretizer.discretize_state(next_state)

            # Update Q-table with experience
            agent.update_q_table(state_discrete, action, reward, next_state_discrete, terminal)

            # Update current state
            state_discrete = next_state_discrete

            # Accumulate episode statistics
            episode_reward += reward
            episode_steps += 1

        # Decay exploration rate
        agent.decay_epsilon()

        # Store episode results
        training_rewards.append(episode_reward)
        training_steps.append(episode_steps)

    return training_rewards, training_steps

### **Implement:** `Cart-Pole Environment`

**Task 5:** **Simulate the environment and display the obtained results.** If you're not using the Gymnasium environment, visualization is not mandatory.

In [200]:
class CartPoleEnvironment:
    """
    Cart-Pole environment that combines physics model, action space, and reward function.
    Provides interface for RL agent to interact with the environment.
    """

    def __init__(self, model, action_space, reward_function):
        """
        Initialize Cart-Pole system Environment.

        Parameters:
            model: CartPoleModel instance for physics simulation
            action_space: ActionSpace instance defining available actions
            reward_function: RewardFunction instance for calculating rewards
        """
        self.model = model
        self.action_space = action_space
        self.reward_function = reward_function

        # Initialize environment state
        self.reset()

    def reset(self):
        """
        Reset environment to initial state with random angle of pole and position of cart.
        """
        self.state = np.array([
            np.random.uniform(-0.05, 0.05),
            0.0,
            np.random.uniform(-0.05, 0.05),
            0.0])

        self.steps_count = 0

        return self.state.copy()

    def step(self, action_idx):
        """
        Execute one step in environment.
        """
        # Get force corresponding passed action index
        force, _ = self.action_space.get_force(action_idx)

        # Get next state with model
        next_state = self.model.next_state(self.state, force)

        # Get reward: For gym_based, no need for theta; otherwise pass next_state[2]
        if self.reward_function.reward_type == 'gym_based':
            reward = self.reward_function.get_reward()  # Fixed +1 for gym_based
        else:
            reward = self.reward_function.get_reward(next_state[2])

        # Update current environment state
        self.state = next_state
        self.steps_count += 1

        # Check if next state is terminal
        terminal = self.is_terminal(next_state, max_steps=500)

        return next_state, reward, terminal

    def is_terminal(self, state, x_limit = 2.4, theta_limit = 0.20944, max_steps = None):
        """
        Determine if current state is terminal (x cross x_limit, theta cross theta_limit, maximum ).
        """
        x, _, theta, _ = state

        if max_steps is None:
            return (abs(x) > x_limit) or (abs(theta) > theta_limit)
        else:
            return (abs(x) > x_limit) or (abs(theta) > theta_limit) or (self.steps_count >= max_steps)

### **Implement:** `Evaluate Trained Q-Learning Agent Method`

In [201]:
def evaluate_agent(environment, state_discretizer, agent, episodes_number, max_steps):
    """
    Evaluate trained agent performance without exploration.
    
    Parameters:
        environment: CartPoleEnvironment instance
        state_discretizer: CartPoleStateDiscretizer instance
        agent: trained QLearningAgent instance
        num_episodes: number of evaluation episodes
        max_steps: maximum steps per episode
    """
    total_rewards = []
    total_steps = []
    succesfull_episodes = 0

    for episode in range(episodes_number):
        # Reset environemnt to initial state
        state = environment.reset()
        state_discrete = state_discretizer.discretize_state(state)

        episode_reward = 0
        episode_steps = 0
        terminal = False

        # Run episode until terminal state or max steps reached
        while not terminal and episode_steps < max_steps:
            # Turn training flag off (False) -> Return action with the best Q-value
            action = agent.get_action(state_discrete, training_active=False)
            next_state, reward, terminal = environment.step(action)
            next_state_discrete = state_discretizer.discretize_state(next_state)

            # Without update of Q-table cuz full exploitation

            state_discrete = next_state_discrete
            episode_reward += reward
            episode_steps += 1

        total_rewards.append(episode_reward)
        total_steps.append(episode_steps)

        # In evaluation of agent successfull episode is one where pole is balanced for 200 steps
        if episode_steps >= max_steps:
            succesfull_episodes += 1

    return total_rewards, total_steps, succesfull_episodes

### **Implement:** `Show Results and Helper Methods`

In [154]:
def plot_results(training_rewards, training_steps, evaluation_rewards, evaluation_steps):

    figure, axes = plt.subplots(2, 2, figsize=(12, 8))

    # 1. Training rewards
    axes[0, 0].plot(training_rewards)
    axes[0, 0].set_title("Training Rewards")
    axes[0, 0].set_xlabel("Episode")
    axes[0, 0].set_ylabel("Total Reward")

    # 2. Training steps
    axes[0, 1].plot(training_steps)
    axes[0, 1].set_title("Training Steps")
    axes[0, 1].set_xlabel("Episode")
    axes[0, 1].set_ylabel("Steps")

    # 3. Evaluation rewards
    axes[1, 0].plot(evaluation_rewards)
    axes[1, 0].set_title("Evaluation Rewards")
    axes[1, 0].set_xlabel("Episode")
    axes[1, 0].set_ylabel("Total Reward")

    # 4. Evaluation steps
    axes[1, 1].plot(evaluation_steps)
    axes[1, 1].set_title("Evaluation Steps")
    axes[1, 1].set_xlabel("Episode")
    axes[1, 1].set_ylabel("Steps")

    # Adjust padding between plots and show
    plt.tight_layout()
    plt.show()


In [155]:
def print_results(rewards, steps, episodes_num,  evaluate_flag = False, successfull_num = 0):

    print("Statistics:")
    print(f"Average reward per episode:     {np.mean(rewards):.2f}")
    print(f"Average number of steps:        {np.mean(steps):.2f}")
    print(f"Min reward per episode:         {min(rewards):.2f}")
    print(f"Min number of steps:            {min(steps):.2f}")
    print(f"Max reward per episode:         {max(rewards):.2f}")
    print(f"Max number of steps:            {max(steps):.2f}")
    
    if evaluate_flag:
        print(f"Number of successfull episodes: {successfull_num:.2f}")
        success_rate = (successfull_num / episodes_num) * 100
        print(f"Success rate:                   {success_rate:.2f}%")

In [156]:
def run_agent(environment, state_discretizer, agent, episodes_num, max_steps):

    print("Q-Learning Agent: Training starts ...")

    training_rewards, training_steps = train_agent(environment = environment,
                                                   state_discretizer = state_discretizer,
                                                   agent = agent,
                                                   episodes_number = episodes_num,
                                                   max_steps = max_steps)

    print("Q-Learning Agent: Training finished.")

    print("Q-Learning Agent: Evaluation starts ...")

    evaluation_rewards, evaluation_steps, successfull_num = evaluate_agent(environment = environment,
                                                                           state_discretizer = state_discretizer,
                                                                           agent = agent,
                                                                           episodes_number = episodes_num,
                                                                           max_steps = max_steps)
    print("Q-Learning Agent: Evaluation finished.")

    plot_results(training_rewards = training_rewards,
                 training_steps = training_steps,
                 evaluation_rewards = evaluation_rewards,
                 evaluation_steps = evaluation_steps)
    
    print("AGENT TRAINING")
    print_results(rewards = training_rewards,
                  steps = training_steps,
                  episodes_num = episodes_num, 
                  evaluate_flag = False,
                  successfull_num = 0)

    print()

    print("AGENT EVALUATION")
    print_results(rewards = evaluation_rewards,
                  steps = evaluation_steps,
                  episodes_num = episodes_num,
                  evaluate_flag = True,
                  successfull_num = successfull_num)

### **Implement:** `Rank Different Q-Learning Agents`

#### Define constants and hyperparameters

In [ ]:
# Define constants
RANDOM_SEED = 42
FORCE_MAX = 10.0
EPISODES_MAX = 2000
STEPS_MAX = 500
GAMMA = 0.95
ALPHA = 0.1
EPSILON = 1.0
EPSILON_MIN = 0.01
EPSILON_DECAY = 0.999

In [20]:
action_space_options = {
    "binary": {"actions_type": "binary", "max_force": 10.0},
    "ternary": {"actions_type": "ternary", "max_force": 10.0},
    "multi": {"actions_type": "multi", "max_force": 10.0}
}

In [21]:
reward_function_options = {
    "angle_based": {"reward_type": "angle_based", "alpha": 2.0, "terminal_theta": math.pi / 6},
    "binary": {"reward_type": "binary", "terminal_theta": math.pi / 6},
    "gym_based": {"reward_type": "gym_based", "terminal_theta": math.pi / 6}
}

In [68]:
def rank_different_agents(model, state_discretizer, episodes_number = 2000, max_steps = 300):
    ranks = []

    np.random.seed(RANDOM_SEED)
    random.seed()

    for action_name, action_settings in action_space_options.items():
        for reward_name, reward_settings in reward_function_options.items():

            action_space = ActionSpace(**action_settings)
            reward_function = RewardFunction(**reward_settings)
            
            environment = CartPoleEnvironment(model, action_space, reward_function)

            agent = QLearningAgent(action_space.action_num)

            training_rewards, training_steps = train_agent(environment,
                                                           state_discretizer,
                                                           agent,
                                                           episodes_number,
                                                           max_steps)
            
            evaluation_rewards, evaluation_steps, success = evaluate_agent(environment, 
                                                                           state_discretizer,
                                                                           agent, 
                                                                           episodes_number,
                                                                           max_steps)
            
            ranks.append({"actions_type": action_name,
                          "reward_type": reward_name,
                          "success_rate": (success / episodes_number) * 100,
                          "average_steps": np.mean(evaluation_steps),
                          "max_steps": max(evaluation_steps)})
            
    return ranks

In [69]:
# Initialize model component
model = CartPoleModel(T = 0.02)

# Initialize state discretizer component
state_discretizer = CartPoleStateDiscretizer(bins = (6, 6, 6, 6))

In [70]:
ranks = rank_different_agents(model, state_discretizer)

In [71]:
print("AGENT[index]:  ACTION TYPE + REWARD TYPE  ")
print("---------------------------------------------")

for index, agent_result in enumerate(ranks):
    print(f"AGENT[{index + 1}]:        {agent_result['actions_type']}  +  {agent_result['reward_type']}")
    print(f"Success Rate:       {agent_result['success_rate']:.2f}%")
    print(f"Average Steps:      {agent_result['average_steps']:.2f}")
    print(f"Max Steps:          {agent_result['max_steps']}")
    print("---------------------------------------------")

AGENT[index]:  ACTION TYPE + REWARD TYPE  
---------------------------------------------
AGENT[1]:        binary  +  angle_based
Success Rate:       2.70%
Average Steps:      114.02
Max Steps:          200
---------------------------------------------
AGENT[2]:        binary  +  binary
Success Rate:       12.05%
Average Steps:      93.99
Max Steps:          200
---------------------------------------------
AGENT[3]:        binary  +  gym_based
Success Rate:       0.00%
Average Steps:      68.57
Max Steps:          145
---------------------------------------------
AGENT[4]:        ternary  +  angle_based
Success Rate:       0.15%
Average Steps:      85.67
Max Steps:          200
---------------------------------------------
AGENT[5]:        ternary  +  binary
Success Rate:       0.75%
Average Steps:      82.13
Max Steps:          200
---------------------------------------------
AGENT[6]:        ternary  +  gym_based
Success Rate:       0.00%
Average Steps:      47.23
Max Steps:        

### **Implement:** `Main Q-Learning Agent & Fine Tuning`

In [ ]:
config = {
    "random_seed": 42,                          
    "time_step": 0.02,                          
    "bins": (6, 6, 12, 12),
    "actions_type": "binary",
    "reward_type": "gym_based",
    "episodes_train": 5000,                     
    "episodes_eval": 500,
    "max_steps": 500,
    "max_force": 10.0,
    "terminal_theta": 0.20944,
    "gamma": 0.98,                              
    "epsilon": 1.0,
    "epsilon_min": 0.005,
    "epsilon_decay": 0.9998,
    "alpha": 0.2
}


In [202]:
def set_seed(seed = 42):
    random.seed(seed)
    np.random.seed(seed)

In [203]:
def train_and_evaluate_agent(config: dict):
    
    set_seed(seed = config['random_seed'])

    model = CartPoleModel(T = config['time_step'])

    action_space = ActionSpace(actions_type = config['actions_type'],
                               max_force = config['max_force'])

    reward_function = RewardFunction(reward_type = config['reward_type'], alpha = 2.0,
                                     terminal_theta = config['terminal_theta'])

    environment = CartPoleEnvironment(model, action_space, reward_function)

    state_discretizer = CartPoleStateDiscretizer(bins = config['bins'])

    agent = QLearningAgent(action_size = action_space.action_num,
                           gamma = config['gamma'],
                           epsilon = config['epsilon'],
                           epsilon_min = config['epsilon_min'],
                           epsilon_decay = config['epsilon_decay'],
                           alpha = config['alpha'])

    training_rewards, training_steps = train_agent(environment, state_discretizer, agent, 
                                                   episodes_number = config['episodes_train'],
                                                   max_steps = config['max_steps'])
    
    evaluation_rewards, evaluation_steps, success = evaluate_agent(environment, state_discretizer,
                                                                   agent, 
                                                                   episodes_number = config['episodes_eval'],
                                                                   max_steps = config['max_steps'])
    
    return training_rewards, training_steps, evaluation_rewards, evaluation_steps, success

In [204]:
def compare_and_display_results(training_rewards, training_steps, evaluation_rewards, evaluation_steps, success):

    episodes_eval = len(evaluation_rewards)

    print("Performance comparison: TRAINING vs EVALUATION")
    
    print("--------------------------------------------------------------")
    print("| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |")
    print("--------------------------------------------------------------")

    train_avg_reward = np.mean(training_rewards)
    eval_avg_reward = np.mean(evaluation_rewards)
    print(f"| Avg. Reward         | {train_avg_reward:10.2f} | {eval_avg_reward:10.2f} | {eval_avg_reward - train_avg_reward:10.2f} |")

    train_avg_steps = np.mean(training_steps)
    eval_avg_steps = np.mean(evaluation_steps)
    print(f"| Avg. Steps          | {train_avg_steps:10.2f} | {eval_avg_steps:10.2f} | {eval_avg_steps - train_avg_steps:10.2f} |")

    train_max_reward = max(training_rewards)
    eval_max_reward = max(evaluation_rewards)
    print(f"| Max. Reward         | {train_max_reward:10.2f} | {eval_max_reward:10.2f} | {eval_max_reward - train_max_reward:10.2f} |")

    train_max_steps = max(training_steps)
    eval_max_steps = max(evaluation_steps)
    print(f"| Max. Steps          | {train_max_steps:10} | {eval_max_steps:10} | {eval_max_steps - train_max_steps:10} |")

    train_std_reward = np.std(training_rewards)
    eval_std_reward = np.std(evaluation_rewards)
    print(f"| Std. Dev. Reward    | {train_std_reward:10.2f} | {eval_std_reward:10.2f} | {eval_std_reward - train_std_reward:10.2f} |")

    print("--------------------------------------------------------------")
    print(f"| Success Rate (Eval) | {'- X -':>10} | {((success / episodes_eval) * 100):.2f}%     | {'- X -':>10} |")
    print("--------------------------------------------------------------")

#### **Fine Tuning Process**

In [179]:
def run_experiment(config):
    training_rewards, training_steps, evaluation_rewards, evaluation_steps, success_rate = train_and_evaluate_agent(config = config)

    compare_and_display_results(training_rewards, training_steps, evaluation_rewards,
                                evaluation_steps, success_rate)

In [180]:
run_experiment(config = config)

Performance comparison: TRAINING vs EVALUATION
--------------------------------------------------------------
| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
--------------------------------------------------------------
| Avg. Reward         |     223.67 |     441.94 |     218.27 |
| Avg. Steps          |     224.50 |     442.14 |     217.65 |
| Max. Reward         |     500.00 |     500.00 |       0.00 |
| Max. Steps          |        500 |        500 |          0 |
| Std. Dev. Reward    |     156.80 |     127.84 |     -28.96 |
--------------------------------------------------------------
| Success Rate (Eval) |      - X - | 88.15%     |      - X - |
--------------------------------------------------------------


In [208]:
def evaluate_agent_unlimited(environment, state_discretizer, agent, episodes_number=5):
    """
    Evaluira agenta za mali broj epizoda bez ograničenja na maksimalni broj koraka.
    Meri se koliko koraka agent može da izdrži pre pada štapa.
    """
    total_steps = []
    
    print(f"\n=== Evaluacija agenta bez ograničenja koraka ({episodes_number} epizoda) ===")
    
    for episode in range(episodes_number):
        # Reset environment to initial state
        state = environment.reset()
        state_discrete = state_discretizer.discretize_state(state)

        episode_steps = 0
        terminal = False

        # Pokreni epizodu dok ne dođe do pada štapa (bez max_steps ograničenja)
        while not terminal:
            # Agent bira akciju (bez treniranja)
            action = agent.get_action(state_discrete, training_active=False)
            
            # Manuelno izvrši korak bez pozivanja environment.step()
            force, _ = environment.action_space.get_force(action)
            next_state = environment.model.next_state(environment.state, force)
            
            # Dobij nagradu
            if environment.reward_function.reward_type == 'gym_based':
                reward = environment.reward_function.get_reward()
            else:
                reward = environment.reward_function.get_reward(next_state[2])
            
            # Updateuj environment state
            environment.state = next_state
            environment.steps_count += 1
            
            # Proveri terminalnost SAMO na osnovu fizičkih ograničenja (bez max_steps)
            terminal = environment.is_terminal(next_state, max_steps=None)
            
            next_state_discrete = state_discretizer.discretize_state(next_state)
            state_discrete = next_state_discrete
            episode_steps += 1
            
            # Bezbednosno ograničenje da se izbegnu beskonačne petlje
            if episode_steps > 10000:
                print(f"  Epizoda {episode + 1}: Prekinuta na 10000 koraka (bezbednosno ograničenje)")
                break

        total_steps.append(episode_steps)
        print(f"  Epizoda {episode + 1}: {episode_steps} koraka")

    # Statistike
    print("\nStatistike neograničene evaluacije:")
    print(f"Prosečan broj koraka:    {np.mean(total_steps):.2f}")
    print(f"Minimalan broj koraka:   {min(total_steps)}")
    print(f"Maksimalan broj koraka:  {max(total_steps)}")
    print(f"Ukupno koraka:           {sum(total_steps)}")
    
    return total_steps

In [209]:
def run_experiment_extended(config):
    """
    Pokretanje eksperimenta sa dodatnom neograničenom evaluacijom
    """
    training_rewards, training_steps, evaluation_rewards, evaluation_steps, success_rate = train_and_evaluate_agent(config = config)

    compare_and_display_results(training_rewards, training_steps, evaluation_rewards,
                                evaluation_steps, success_rate)
    
    # Kreiraj komponente potrebne za neograničenu evaluaciju
    set_seed(seed = config['random_seed'])
    
    model = CartPoleModel(T = config['time_step'])
    action_space = ActionSpace(actions_type = config['actions_type'], max_force = config['max_force'])
    reward_function = RewardFunction(reward_type = config['reward_type'], alpha = 2.0, terminal_theta = config['terminal_theta'])
    environment = CartPoleEnvironment(model, action_space, reward_function)
    state_discretizer = CartPoleStateDiscretizer(bins = config['bins'])
    
    # Kreiraj agenta sa istim parametrima i ponovo ga istrenira
    agent = QLearningAgent(action_size = action_space.action_num,
                           gamma = config['gamma'],
                           epsilon = config['epsilon'],
                           epsilon_min = config['epsilon_min'],
                           epsilon_decay = config['epsilon_decay'],
                           alpha = config['alpha'])
    
    # Ponovo treniraj agenta
    train_agent(environment, state_discretizer, agent, episodes_number = config['episodes_train'], max_steps = config['max_steps'])
    
    # Pokreni neograničenu evaluaciju
    unlimited_steps = evaluate_agent_unlimited(environment, state_discretizer, agent)
    
    return unlimited_steps

### **Test unlimited steps in evaluation**

In [210]:
config = {
    "random_seed": 42,              # možeš menjati u for-petlji kao što si radio
    "time_step": 0.02,              
    "bins": (10, 10, 18, 18),       # malo finija discretizacija
    "actions_type": "binary",       # binary obično bolje radi
    "reward_type": "angle_based",   # daje više informacija tokom treninga
    "episodes_train": 20000,        # više epizoda za stabilnije učenje
    "episodes_eval": 2000,          # evaluacija na većem uzorku
    "max_steps": 500,
    "max_force": 10.0,
    "terminal_theta": 0.20944,
    "gamma": 0.98,                  # dobar kompromis između dugoročnog i kratkoročnog
    "epsilon": 1.0,
    "epsilon_min": 0.05,            # agent ne sme previše rano da prestane da istražuje
    "epsilon_decay": 0.999,        # brže smanjenje eksploracije nego kod 0.9999
    "alpha": 0.1                    # manji learning rate → stabilnije konvergira
}

unlimited_results = run_experiment_extended(config)

Performance comparison: TRAINING vs EVALUATION
--------------------------------------------------------------
| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
--------------------------------------------------------------
| Avg. Reward         |    -264.08 |     -66.70 |     197.39 |
| Avg. Steps          |     452.66 |     500.00 |      47.34 |
| Max. Reward         |     -42.56 |     -43.96 |      -1.40 |
| Max. Steps          |        500 |        500 |          0 |
| Std. Dev. Reward    |     336.69 |      10.10 |    -326.59 |
--------------------------------------------------------------
| Success Rate (Eval) |      - X - | 100.00%     |      - X - |
--------------------------------------------------------------

=== Evaluacija agenta bez ograničenja koraka (5 epizoda) ===
  Epizoda 1: 2467 koraka
  Epizoda 2: 2420 koraka
  Epizoda 3: 2483 koraka
  Epizoda 4: 2368 koraka
  Epizoda 5: 2426 koraka

Statistike neograničene evaluacije:
Prosečan broj koraka:    2432.80
Mi

In [212]:
config = {
    "random_seed": 42,              # možeš menjati u for-petlji kao što si radio
    "time_step": 0.02,              
    "bins": (12, 12, 24, 24),       # malo finija discretizacija
    "actions_type": "binary",       # binary obično bolje radi
    "reward_type": "angle_based",   # daje više informacija tokom treninga
    "episodes_train": 20000,        # više epizoda za stabilnije učenje
    "episodes_eval": 2000,          # evaluacija na većem uzorku
    "max_steps": 500,
    "max_force": 10.0,
    "terminal_theta": 0.20944,
    "gamma": 0.98,                  # dobar kompromis između dugoročnog i kratkoročnog
    "epsilon": 1.0,
    "epsilon_min": 0.05,            # agent ne sme previše rano da prestane da istražuje
    "epsilon_decay": 0.999,        # brže smanjenje eksploracije nego kod 0.9999
    "alpha": 0.1                    # manji learning rate → stabilnije konvergira
}

unlimited_results = run_experiment_extended(config)

Performance comparison: TRAINING vs EVALUATION
--------------------------------------------------------------
| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
--------------------------------------------------------------
| Avg. Reward         |    -300.83 |     -72.64 |     228.19 |
| Avg. Steps          |     437.61 |     500.00 |      62.39 |
| Max. Reward         |     -31.65 |     -42.96 |     -11.31 |
| Max. Steps          |        500 |        500 |          0 |
| Std. Dev. Reward    |     362.19 |      11.17 |    -351.02 |
--------------------------------------------------------------
| Success Rate (Eval) |      - X - | 100.00%     |      - X - |
--------------------------------------------------------------

=== Evaluacija agenta bez ograničenja koraka (5 epizoda) ===
  Epizoda 1: 689 koraka
  Epizoda 2: 873 koraka
  Epizoda 3: 812 koraka
  Epizoda 4: 740 koraka
  Epizoda 5: 726 koraka

Statistike neograničene evaluacije:
Prosečan broj koraka:    768.00
Minimala

### **Implement:** `Cart-Pole SARSA Agent and Training`

In [188]:
class SARSAAgent:
    """
    SARSA agent for solving Cart-Pole problem (on-policy TD control).
    """

    def __init__(self, action_size, gamma=0.95, epsilon=1.0, epsilon_min=0.01,
                 epsilon_decay=0.995, alpha=0.1):
        self.action_size = action_size
        self.gamma = gamma
        self.epsilon = epsilon
        self.epsilon_min = epsilon_min
        self.epsilon_decay = epsilon_decay
        self.alpha = alpha

        self.q_table = defaultdict(lambda: np.zeros(action_size))

    def get_action(self, state, training_active=True):
        if training_active and np.random.random() < self.epsilon:
            return np.random.randint(0, self.action_size)
        else:
            return np.argmax(self.q_table[state])

    def decay_epsilon(self):
        self.epsilon = max(self.epsilon_min, self.epsilon * self.epsilon_decay)

    def update_q_table(self, state, action, reward, next_state, next_action, terminal):
        """
        SARSA update rule:
        Q(s,a) <- Q(s,a) + alpha * (r + gamma*Q(s',a') - Q(s,a))
        """
        if terminal:
            td_target = reward
        else:
            td_target = reward + self.gamma * self.q_table[next_state][next_action]

        td_error = td_target - self.q_table[state][action]
        self.q_table[state][action] += self.alpha * td_error


In [189]:
def train_agent_sarsa(environment, state_discretizer, agent, episodes_number, max_steps=500):
    training_rewards = []
    training_steps = []

    for episode in range(episodes_number):
        state = environment.reset()
        state_discrete = state_discretizer.discretize_state(state)

        action = agent.get_action(state_discrete)
        episode_reward = 0
        episode_steps = 0
        terminal = False

        while not terminal and episode_steps < max_steps:
            next_state, reward, terminal = environment.step(action)
            next_state_discrete = state_discretizer.discretize_state(next_state)

            next_action = agent.get_action(next_state_discrete)

            # Update Q using SARSA rule
            agent.update_q_table(state_discrete, action, reward, next_state_discrete, next_action, terminal)

            state_discrete = next_state_discrete
            action = next_action

            episode_reward += reward
            episode_steps += 1

        agent.decay_epsilon()
        training_rewards.append(episode_reward)
        training_steps.append(episode_steps)

    return training_rewards, training_steps


In [190]:
def train_and_evaluate_SARSA_agent(config: dict):
    
    set_seed(seed = config['random_seed'])

    model = CartPoleModel(T = config['time_step'])

    action_space = ActionSpace(actions_type = config['actions_type'],
                               max_force = config['max_force'])

    reward_function = RewardFunction(reward_type = config['reward_type'], alpha = 2.0,
                                     terminal_theta = config['terminal_theta'])

    environment = CartPoleEnvironment(model, action_space, reward_function)

    state_discretizer = CartPoleStateDiscretizer(bins = config['bins'])

    agent = SARSAAgent(
        action_size=action_space.action_num,
        gamma=config['gamma'],
        epsilon=config['epsilon'],
        epsilon_min=config['epsilon_min'],
        epsilon_decay=config['epsilon_decay'],
        alpha=config['alpha']
    )

    training_rewards, training_steps = train_agent_sarsa(environment, state_discretizer, agent,
                                                        episodes_number=config['episodes_train'],
                                                        max_steps=config['max_steps'])

    evaluation_rewards, evaluation_steps, success = evaluate_agent(environment, state_discretizer,
                                                                   agent, 
                                                                   episodes_number = config['episodes_eval'],
                                                                   max_steps = config['max_steps'])
    
    return training_rewards, training_steps, evaluation_rewards, evaluation_steps, success

In [191]:
def run_sarsa_experiment(config):
    training_rewards, training_steps, evaluation_rewards, evaluation_steps, success_rate = train_and_evaluate_SARSA_agent(config = config)

    compare_and_display_results(training_rewards, training_steps, evaluation_rewards,
                                evaluation_steps, success_rate)

In [192]:
config = {
    "random_seed": 42,
    "time_step": 0.02,
    "bins": (8, 8, 16, 16), 
    "actions_type": "binary",
    "reward_type": "angle_based",
    "episodes_train": 30000,
    "episodes_eval": 1000,
    "max_steps": 500,
    "max_force": 10.0,
    "terminal_theta": 0.20944,
    "gamma": 0.95,
    "epsilon": 1.0,
    "epsilon_min": 0.01,
    "epsilon_decay": 0.9999,
    "alpha": 0.1
}

In [193]:
run_sarsa_experiment(config = config)

Performance comparison: TRAINING vs EVALUATION
--------------------------------------------------------------
| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
--------------------------------------------------------------
| Avg. Reward         |    -654.48 |     -55.53 |     598.95 |
| Avg. Steps          |     303.75 |     500.00 |     196.25 |
| Max. Reward         |     -44.52 |     -36.50 |       8.02 |
| Max. Steps          |        500 |        500 |          0 |
| Std. Dev. Reward    |     511.75 |       9.74 |    -502.01 |
--------------------------------------------------------------
| Success Rate (Eval) |      - X - | 100.00%     |      - X - |
--------------------------------------------------------------


In [194]:
for seed in [0, 1, 2, 7, 13, 21, 42, 99, 123, 2025]:
    config["random_seed"] = seed
    print(f"\n===== SEED = {seed} =====")
    run_sarsa_experiment(config)


===== SEED = 0 =====
Performance comparison: TRAINING vs EVALUATION
--------------------------------------------------------------
| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
--------------------------------------------------------------
| Avg. Reward         |    -654.97 |    -117.63 |     537.34 |
| Avg. Steps          |     303.43 |     478.93 |     175.50 |
| Max. Reward         |     -35.41 |     -37.50 |      -2.09 |
| Max. Steps          |        500 |        500 |          0 |
| Std. Dev. Reward    |     514.14 |      71.63 |    -442.51 |
--------------------------------------------------------------
| Success Rate (Eval) |      - X - | 100.00%     |      - X - |
--------------------------------------------------------------

===== SEED = 1 =====
Performance comparison: TRAINING vs EVALUATION
--------------------------------------------------------------
| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
------------------------------------------

### **Cart Pole Q-Learning - Results**

#### 1.0 Initial Results (Before Modifications)

| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
|---------------------|------------|------------|------------|
| Avg. Reward         | -1927.95   | -2.80      | 1925.15    |
| Avg. Steps          | 76.64      | 90.95      | 14.30      |
| Max. Reward         | 0.00       | 0.00       | 0.00       |
| Max. Steps          | 200        | 117        | -83        |
| Std. Dev. Reward    | 1257.44    | 32.28      | -1225.16   |
| Success Rate (Eval) | - X -      | 0.00%      | - X -      |

#### 1.1 Modifying Initial Values in the Reset Function

**Change:** Adjusted the bins during environment reset.  
Previously, the initial angle used values from the range -0.5 to 0.5 (≈ 28 degrees), which was too large.  
Now, each state variable receives a randomized value in the range -0.01 to 0.01.  

| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
|---------------------|------------|------------|------------|
| Avg. Reward         | -1882.50   | -234.95    | 1647.55    |
| Avg. Steps          | 76.83      | 101.16     | 24.33      |
| Max. Reward         | 0.00       | 0.00       | 0.00       |
| Max. Steps          | 200        | 129        | -71        |
| Std. Dev. Reward    | 1234.34    | 423.44     | -810.90    |
| Success Rate (Eval) | - X -      | 0.00%      | - X -      |



#### 1.2 & 1.3 Terminal State and Bin Adjustments

**Changes:**  
- The terminal state check previously used a fixed limit of 90 degrees and a hard-coded maximum number of steps. Now, both the angle limit and maximum episode length can be customized.  
- Finer bins were introduced for improved state discretization.  


| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
|---------------------|------------|------------|------------|
| Avg. Reward         | -85.75     | -0.35      | 85.40      |
| Avg. Steps          | 111.37     | 155.92     | 44.55      |
| Max. Reward         | 0.00       | 0.00       | 0.00       |
| Max. Steps          | 500        | 428        | -72        |
| Std. Dev. Reward    | 34.96      | 5.91       | -29.05     |
| Success Rate (Eval) | - X -      | 4.70%      | - X -      |



#### 1.4 Switching to Gym-Based Reward Function

**Change:** Replaced the previous reward function with the Gym-based reward scheme.  

| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
|---------------------|------------|------------|------------|
| Avg. Reward         | 34.31      | 27.80      | -6.52      |
| Avg. Steps          | 35.31      | 28.80      | -6.52      |
| Max. Reward         | 144.00     | 155.00     | 11.00      |
| Max. Steps          | 145        | 156        | 11         |
| Std. Dev. Reward    | 19.99      | 14.56      | -5.44      |
| Success Rate (Eval) | - X -      | 0.00%      | - X -      |


#### 1.5 Returning to Binary Reward with Higher Learning Rate

**Change:** Reverted to Binary Reward Function. Increased learning rate α from 0.1 to 0.2.  

| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
|---------------------|------------|------------|------------|
| Avg. Reward         | -87.00     | -10.65     | 76.35      |
| Avg. Steps          | 104.32     | 206.38     | 102.07     |
| Max. Reward         | 0.00       | 0.00       | 0.00       |
| Max. Steps          | 500        | 419        | -81        |
| Std. Dev. Reward    | 33.63      | 30.85      | -2.78      |
| Success Rate (Eval) | - X -      | 45.25%     | - X -      |

#### 1.6 Increasing Number of Training Episodes

**Change:** Increased training episodes from 2000 to 5000 (evaluation kept at 2000).  

| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
|---------------------|------------|------------|------------|
| Avg. Reward         | -44.86     | -0.05      | 44.81      |
| Avg. Steps          | 227.16     | 379.58     | 152.42     |
| Max. Reward         | 0.00       | 0.00       | 0.00       |
| Max. Steps          | 500        | 500        | 0          |
| Std. Dev. Reward    | 49.74      | 2.24       | -47.50     |
| Success Rate (Eval) | - X -      | 89.55%     | - X -      |


#### 1.7 Lowering the Terminal Theta Parameter

**Change:** Adjusted terminal theta in the reward function from 90° (π/2) to 12° (~0.21 radians).  

| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
|---------------------|------------|------------|------------|
| Avg. Reward         | -44.86     | -0.05      | 44.81      |
| Avg. Steps          | 227.16     | 379.58     | 152.42     |
| Max. Reward         | 0.00       | 0.00       | 0.00       |
| Max. Steps          | 500        | 500        | 0          |
| Std. Dev. Reward    | 49.74      | 2.24       | -47.50     |
| Success Rate (Eval) | - X -      | 89.55%     | - X -      |

#### 1.8 Q-Learning Agent Check after Environment Fix 

Bootstrap for terminal states was removed.  

| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
|---------------------|------------|------------|------------|
| Avg. Reward         | -42.46     | -1.40      | 41.06      |
| Avg. Steps          | 144.82     | 201.70     | 56.88      |
| Max. Reward         | 0.00       | 0.00       | 0.00       |
| Max. Steps          | 500        | 309        | -191       |
| Std. Dev. Reward    | 49.43      | 11.75      | -37.68     |
| Success Rate (Eval) | - X -      | 48.85%     | - X -      |

#### 1.9 Denominator Check Update 

 **Change:** Replaced denominator check (`!= 0`) with a more stable condition: greater than `1e-10`.  

| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
|---------------------|------------|------------|------------|
| Avg. Reward         | -30.80     | 0.00       | 30.80      |
| Avg. Steps          | 88.48      | 109.25     | 20.77      |
| Max. Reward         | 0.00       | 0.00       | 0.00       |
| Max. Steps          | 130        | 118        | -12        |
| Std. Dev. Reward    | 46.17      | 0.00       | -46.17     |
| Success Rate (Eval) | - X -      | 0.00%      | - X -      |

#### 1.10 State Reset Range Adjustment

**Change:** State reset range increased from ±0.05 to ±0.1.  

| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
|---------------------|------------|------------|------------|
| Avg. Reward         | -31.58     | -0.75      | 30.83      |
| Avg. Steps          | 87.56      | 108.94     | 21.38      |
| Max. Reward         | 0.00       | 0.00       | 0.00       |
| Max. Steps          | 128        | 126        | -2         |
| Std. Dev. Reward    | 46.48      | 8.63       | -37.86     |
| Success Rate (Eval) | - X -      | 0.00%      | - X -      |

#### 1.11 Reset Back to ±0.05, Updated State Discretizer

**Change:** Reverted reset range to ±0.05. Modified `CartPoleStateDiscretizer` to fix bin boundary calculations.  

**Observation:** The agent managed to reach maximum episode length, confirming that the previous mathematical formulas and their implementation were flawed.  

| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
|---------------------|------------|------------|------------|
| Avg. Reward         | -64.76     | -0.60      | 64.16      |
| Avg. Steps          | 123.67     | 190.85     | 67.19      |
| Max. Reward         | 0.00       | 0.00       | 0.00       |
| Max. Steps          | 500        | 500        | 0          |
| Std. Dev. Reward    | 47.77      | 7.72       | -40.05     |
| Success Rate (Eval) | - X -      | 29.90%     | - X -      |

#### 1.12 Gym-Based Reward (Revisited)

| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
|---------------------|------------|------------|------------|
| Avg. Reward         | 155.57     | 162.57     | 7.00       |
| Avg. Steps          | 156.51     | 163.57     | 7.06       |
| Max. Reward         | 500.00     | 500.00     | 0.00       |
| Max. Steps          | 500        | 500        | 0          |
| Std. Dev. Reward    | 130.00     | 53.64      | -76.36     |
| Success Rate (Eval) | - X -      | 16.40%     | - X -      |

#### 1.13 Training Episodes Increased from 5000 to 10000

| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
|---------------------|------------|------------|------------|
| Avg. Reward         | 203.95     | 196.43     | -7.52      |
| Avg. Steps          | 204.80     | 197.37     | -7.44      |
| Max. Reward         | 500.00     | 500.00     | 0.00       |
| Max. Steps          | 500        | 500        | 0          |
| Std. Dev. Reward    | 154.38     | 125.06     | -29.32     |
| Success Rate (Eval) | - X -      | 33.90%     | - X -      |

#### 1.14 Training Episodes Increased from 10000 to 20000

| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
|---------------------|------------|------------|------------|
| Avg. Reward         | 223.67     | 441.94     | 218.27     |
| Avg. Steps          | 224.50     | 442.14     | 217.65     |
| Max. Reward         | 500.00     | 500.00     | 0.00       |
| Max. Steps          | 500        | 500        | 0          |
| Std. Dev. Reward    | 156.80     | 127.84     | -28.96     |
| Success Rate (Eval) | - X -      | 88.15%     | - X -      |

#### 1.15 Evaluation Success Criteria Update  

**Change:** The maximum number of steps per episode was adjusted from 200 to 500.  
Additionally, the Gym-based reward function was updated to match Gymnasium: each survived step yields a reward of +1.  

| METRIC              | TRAINING   | EVALUATION | DIFFERENCE |
|---------------------|------------|------------|------------|
| Avg. Reward         | 223.67     | 441.94     | 218.27     |
| Avg. Steps          | 224.50     | 442.14     | 217.65     |
| Max. Reward         | 500.00     | 500.00     | 0.00       |
| Max. Steps          | 500        | 500        | 0          |
| Std. Dev. Reward    | 156.80     | 127.84     | -28.96     |
| Success Rate (Eval) | - X -      | 88.15%     | - X -      |